src/backend/distributed/utils/metadata_cache.c

```diff
@@ -2731,6 +2723,22 @@ InvalidateDistRelationCacheCallback(Datum argument, Oid relationId)
 }
 
 
+/* InvalidateDistRelationCache flushes the entire distributed relation cache. */
+void
+InvalidateDistRelationCache(void)
+{
+	DistTableCacheEntry *cacheEntry = NULL;
+	HASH_SEQ_STATUS status;
+
+	hash_seq_init(&status, DistTableCacheHash);
+
+	while ((cacheEntry = (DistTableCacheEntry *) hash_seq_search(&status)) != NULL)
+	{
+		cacheEntry->isValid = false;
```

Comment:
There's a slight problem here that we never actually clean entries of dropped relations, which creates a memory leak. We've introduced these flags because invalidations can arrive in lots of different places, and callers might still be using the cache entries, so it wasn't safe to immediately free them. However, since we're calling this code explicitly before we do anything else, and only do this once or twice a day, maybe we should free the entries and do a hash_destroy.

(ideally, we wouldn't use the cache)

Reply:
Note that this leak also exists for regular sessions, but we found this an acceptable trade-off because sessions are ultimately short-lived.

Reply:
Done.